### DISCUSSION TITLE
Allow Jobs to run in sequence

### DISCUSSION THREAD

**Discussion by amitaibu** [digitallyinduced/ihp#1143] [Ideas]
I have a demo "Seat reservation" app that allows students to reserve a seat.

When students asks for a seats, a Reservation record is saved, and a unique seat number should be later assigned to it via a Job.
[See](https://obcbzjejwjytmrlhiepnpklaobjjhpof.ihpapp.com/ShowLibraryOpening?libraryOpeningId=f2d102de-fe4d-4c21-9748-2219e095aa34) it in action.

When I try to register many students at once, I get duplicate seat numbers.

```
time seq 20 | parallel -n0 "curl 'https://obcbzjejwjytmrlhiepnpklaobjjhpof.ihpapp.com/CreateReservation' -H 'content-type:
application/x-www-form-urlencoded' --data-raw 'libraryOpeningId=f2d102de-fe4d-4c21-9748-2219e095aa34&studentIdentifier=1234' --compressed"
```

![Selection_999(338)](https://user-images.githubusercontent.com/125707/137429606-2b01d764-9f5f-45bf-bd5a-816fcf3a3375.jpg)

It would have been great if on a Jobs definition I could define

```haskell
instance Job ReservationJob where
    runMultipleJobsInParallel = False
````

**Comment by amitaibu** [comment]
Currently in order to prevent the multiple seats I'm using

```haskell
perform ReservationJob { .. } = withTransaction do
    sqlExec "LOCK TABLE reservations" ()
```

So Maybe that's already doing the sequence I require.

**Reply by amitaibu** [reply]
Actually, that still doesn't work.

**Reply by mpscholten** [reply]
Maybe you could use a postgresql sequence for this? Basically creating a sequence for every library opening (using `sqlQuery` and `sqlExec` as there's no functionality for sequences in IHP itself). Than you could use `nextval()` inside a `sqlExec` call to assign a unique seat number

**Reply by amitaibu** [reply]
You mean a sequence for the seat number? If so, It's actually more involved than that, above example is a simplified version (in rela life student may choose different areas to seat, and we should find a matching seat number). The non-parallel is also for my Bid system proof of concept - In both cases I need a way to process items one of the other, and prevent race conditions, as those requests tend to come in parallel.

**Reply by amitaibu** [reply]
As an aside, if I'll be able to solve that logic for the Seat reservation, I might be able to turn it into a paid client work :crossed_fingers:

**Reply by mpscholten** [reply]
Ok, i think if sequences are not an option we could definitly add a `runMultipleJobsInParallel` option 👍

**Reply by mpscholten** [reply]
Try out https://github.com/digitallyinduced/ihp/pull/1144 It allows to set a `maxConcurrency` on the job. Setting it to `maxConcurrency = 1` should solve this problem

**Reply by amitaibu** [reply]
Wow, thank you for working on that @mpscholten ! Going to try and learn from the PR code.

I notice there are some errors (also on GitHub Actions)

![Selection_999(340)](https://user-images.githubusercontent.com/125707/137502917-ebd68a19-a46b-4a9e-8d0e-7ad6b92d9eec.jpg)

**Reply by mpscholten** [reply]
Sorry, broke master as well :) Fixed it now

**Reply by amitaibu** [reply]
Seems also master is still red

**Reply by mpscholten** [reply]
Another error :) fixed now

**Reply by amitaibu** [reply]
Thanks. It seems as if the queue is no longer processing. I've re-created DB and re-deployed on IHPcloud, but queue items are not processed (same on my local)

https://obcbzjejwjytmrlhiepnpklaobjjhpof.ihpapp.com/ShowLibraryOpening?libraryOpeningId=b7eba01b-88c9-4361-ae61-35e3b9475b88

On my local logs I see: `NOTICE:  trigger "did_insert_job_reservation_jobs" for relation "reservation_jobs" does not exist, skipping`

**Reply by mpscholten** [reply]
Ah yes, the `withTaskGroup` is already stopped once it's callback function has returned. Therefore most tasks were directly killed after startup. Fixed this now on master

**Reply by amitaibu** [reply]
It works :partying_face:
